src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -190,7 +192,40 @@ class SchemaService(private val options: Neo4jOptions, private val driverCache:
     val query = queryReadStrategy.createStatementForQuery(options)
     val params = Collections.singletonMap[String, AnyRef](Neo4jQueryStrategy.VARIABLE_SCRIPT_RESULT, Collections.emptyList())
     val structFields = retrieveSchema(query, params, { record => record.asMap.asScala.toMap })
-    StructType(structFields)
+
+    val result = session.run(s"EXPLAIN $query").consume()
+    val plan = result.plan()
+
+    val columns = if (plan.arguments().containsKey("Details")) {
+      plan.arguments()
+        .get("Details")
+        .asString()
+        .replaceAll("\"", "")
+        .split(',')
+        .map(_.trim)
+    }
+    else {
+      val lastChild = plan.children().get(0)
+
+      lastChild.operatorType() match {
+        case "EagerAggregation" => lastChild.identifiers().asScala.toArray
+        case "ProcedureCall" => plan.identifiers().asScala.toArray
+        case _ =>
+          val expressions = lastChild.arguments().get("Expressions").asString()
+
+          val firstLevelExpressions = "\\{(.*?)}".r.replaceAllIn(expressions.substring(1,expressions.length-1),"_")
+
+          "([^,:]*?):".r.findAllMatchIn(firstLevelExpressions).map(_.group(1).trim).toArray
+      }
+    }
+
+    val sortedStructFields = if (structFields.isEmpty) {
+      columns.map(StructField(_, DataTypes.StringType))
```

Comment:
please add a comment highlighting that we want to return an empty df with the requested columns